File: src/main/java/com/offz/minecraft/staminaclimbing/plugin/Stamina/StaminaTask.java
package com.offz.minecraft.staminaclimbing.plugin.Stamina;

import com.offz.minecraft.staminaclimbing.plugin.Climbing.ClimbBehaviour;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.util.Vector;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;
import java.util.UUID;

public class StaminaTask extends BukkitRunnable {
    @Override
    public void run() {
        for (UUID uuid : StaminaBar.registeredBars.keySet()) { //Regenerate stamina for every BossBar
            Player p = Bukkit.getPlayer(uuid);
            BossBar b = StaminaBar.registeredBars.get(uuid);
            double progress = b.getProgress();

            if (progress + 0.01 <= 1) {
                b.setVisible(true);
                b.setProgress(progress + 0.01);
            } else {
                b.setVisible(false);
                b.setProgress(1);
            }

            if (progress <= 0.02) { //Changing bar colors and effects on player depending on its progress
                b.setColor(BarColor.RED);
                b.setTitle(ChatColor.translateAlternateColorCodes('&', "&c&lStamina")); //Make Stamina title red
                if (ClimbBehaviour.isClimbing.containsKey(uuid)) {
//                    p.removePotionEffect(PotionEffectType.LEVITATION);
                    ClimbBehaviour.stopClimbing(p);
                }

                ClimbBehaviour.canClimb.put(uuid, false); //If player reaches red zone, they can't climb until they get back in green zone
//                p.setSaturation(0);
                p.addPotionEffect(new PotionEffect(PotionEffectType.SLOW, 110, 2, false, false));
                p.addPotionEffect(new PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false));
//                if (ClimbBehaviour.isClimbing.containsKey(uuid) && ClimbBehaviour.isClimbing.remove(uuid))
//                    p.removePotionEffect(PotionEffectType.LEVITATION);
            } else if (progress < 1) {
                if (!ClimbBehaviour.canClimb.get(uuid)) { //Keep Stamina Bar red even in yellow zone while it's regenerating
                    b.setColor(BarColor.RED);
                    continue;
                }
            } else {
                b.setColor(BarColor.GREEN);
                b.setTitle(ChatColor.BOLD + "Stamina");
                ClimbBehaviour.canClimb.put(uuid, true);
            }
        }
        for (UUID uuid : ClimbBehaviour.isClimbing.keySet()) {
            Player p = Bukkit.getPlayer(uuid);
            Location loc = p.getLocation();
            Vector direction = p.getLocation().getDirection();
            Vector v = p.getVelocity();

            boolean isClimbing = ClimbBehaviour.isClimbing.get(uuid);
//            if(isClimbing && ClimbBehaviour.cooldown.get(uuid) <= System.currentTimeMillis()) {
//            }



            if (!p.isFlying() && isClimbing) {
                ClimbBehaviour.stopClimbing(p);
                continue;
            }

            if (ClimbBehaviour.atWall(p.getLocation())) {
                if (ClimbBehaviour.cooldown.get(uuid) <= System.currentTimeMillis())
                    ClimbBehaviour.cooldown.put(uuid, System.currentTimeMillis());
                p.setAllowFlight(true);
                p.setFlying(true);
                ClimbBehaviour.isClimbing.put(uuid, true);
            } else {
                if (isClimbing) {
                    p.setVelocity(new Vector(v.getX() + direction.getX() / 20, v.getY(), v.getZ() + direction.getZ() / 20));
                    ClimbBehaviour.isClimbing.put(uuid, false);
                    ClimbBehaviour.cooldown.put(uuid, System.currentTimeMillis());
                    p.setFlying(false);
                    p.setAllowFlight(false);
                } else if (ClimbBehaviour.cooldown.get(uuid) + 400 <= System.currentTimeMillis()) {
                    ClimbBehaviour.stopClimbing(p);
                    continue;
                }
            }

            isClimbing = ClimbBehaviour.isClimbing.get(uuid);

            /*if ((p.isSneaking() || loc.add(0, -0.2, 0).getBlock().getType().isSolid()) && ClimbBehaviour.cooldown.get(uuid) + 160 <= System.currentTimeMillis()) {
                ClimbBehaviour.stopClimbing(p);
                continue;
            }*/

            /*if (ClimbBehaviour.cooldown.get(uuid) + 50 <= System.currentTimeMillis()) {
                p.removePotionEffect(PotionEffectType.LEVITATION);
                p.addPotionEffect(new PotionEffect(PotionEffectType.LEVITATION, Integer.MAX_VALUE, 255, false, false));
            }*/

//            if(Math.abs(v.getX()) > 0.05 || Math.abs(v.getY()) > 0.05 || Math.abs(v.getZ()) > 0.05)
//                p.sendMessage("Using");
//                StaminaBar.removeProgress(0.01, b);

            StaminaBar.removeProgress(0.014, uuid);
        }
    }
}

File: src/main/kotlin/com/mineinabyss/staminaclimb/stamina/StaminaTask.kt
package com.mineinabyss.staminaclimb.stamina

import com.mineinabyss.geary.papermc.tracking.items.inventory.toGeary
import com.mineinabyss.idofront.entities.toPlayer
import com.mineinabyss.idofront.location.down
import com.mineinabyss.idofront.textcomponents.miniMsg
import com.mineinabyss.idofront.time.inWholeTicks
import com.mineinabyss.staminaclimb.*
import com.mineinabyss.staminaclimb.climbing.ClimbBehaviour
import com.mineinabyss.staminaclimb.config.StaminaConfig
import net.kyori.adventure.bossbar.BossBar
import org.bukkit.GameMode
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.scheduler.BukkitRunnable
import kotlin.time.Duration.Companion.nanoseconds

class StaminaTask(private val config: StaminaConfig) : BukkitRunnable() {
    private var lastTickNano = System.nanoTime()
    private var timeSinceLastColorFlip = 0L
    private var lastTime = System.currentTimeMillis()

    private val redBar = "<red><b>Stamina".miniMsg()
    private val baseBar = "<b>Stamina".miniMsg()

    override fun run() {
        val currentNano = System.nanoTime()
        val tickDuration = (currentNano - lastTickNano).nanoseconds.inWholeTicks
        lastTickNano = currentNano

        StaminaBar.forEachBar { player, uuid, bar ->
            val progress = bar.progress()

            if (player.gameMode == GameMode.CREATIVE || player.gameMode == GameMode.SPECTATOR) {
                StaminaBar.unregisterBar(uuid)
                player.stopClimbing()
                player.allowFlight = true
                return
            }

            //regenerate stamina for BossBar
            if (!uuid.isClimbing)
                bar.addProgress(
                    when {
                        player.location.down(0.0625).block.isSolid -> config.staminaRegen
                        !player.isInClimbableBlock -> config.staminaRegenInAir
                        else -> 0f
                    }
                )

            when {
                progress <= config.barRed -> { //Changing bar colors and effects on player depending on its progress
                    bar.color(BossBar.Color.RED)
                    bar.name(redBar)
                    if (uuid.isClimbing) player.stopClimbing()

                    uuid.canClimb = false//If player is in red zone, they can't climb until they get back in green zone
                    player.addPotionEffect(PotionEffect(PotionEffectType.SLOWNESS, 110, 2, false, false))
                    player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
                }

                progress < 1 && !uuid.canClimb -> bar.color(BossBar.Color.RED) //Keep Stamina Bar red even in yellow zone while it's regenerating
                (uuid.isClimbing || player.isInClimbableBlock) && progress <= config.barBlink2 -> {
                    val deltaTime = System.currentTimeMillis() - lastTime
                    lastTime = System.currentTimeMillis()
                    if (timeSinceLastColorFlip < config.barBlinkSpeed2)
                        timeSinceLastColorFlip += deltaTime
                    else {
                        flipColor(bar)
                        timeSinceLastColorFlip = 0
                    }
                }

                (uuid.isClimbing || player.isInClimbableBlock) && progress <= config.barBlink1 -> {
                    val deltaTime = System.currentTimeMillis() - lastTime
                    lastTime = System.currentTimeMillis()
                    if (timeSinceLastColorFlip < config.barBlinkSpeed1)
                        timeSinceLastColorFlip += deltaTime
                    else {
                        flipColor(bar)
                        timeSinceLastColorFlip = 0
                    }
                }

                else -> {
                    bar.color(config.baseBarColor)
                    bar.name(baseBar)
                    uuid.canClimb = true
                }
            }
        }

        ClimbBehaviour.isClimbing.entries.forEach { (uuid, isClimbing) ->
            val player = uuid.toPlayer() ?: ClimbBehaviour.isClimbing.remove(uuid).let { return }
            //if climbing in creative, stop climbing but keep flight
            if (player.gameMode == GameMode.CREATIVE) {
                player.stopClimbing()
                player.allowFlight = true
                player.isFlying = true
                return@forEach
            }

            //prevent player from climbing if they have fallen far enough or in a invalid state
            if (!player.isFlying && isClimbing || player.fallDistance > config.maxFallDist) {
                player.stopClimbing()
                return@forEach
            }

            val atWallMultiplier = player.wallDifficulty
            if (atWallMultiplier >= 0) {
                if (uuid.climbCooldownDone) uuid.restartCooldown()
                player.allowFlight = true
                player.isFlying = true
                ClimbBehaviour.isClimbing[uuid] = true
            } else {
                if (isClimbing) {
                    uuid.isClimbing = false
                    uuid.restartCooldown()
                    player.launchInDirection()
                    player.isFlying = false
                    player.allowFlight = false
                }
                //only prevent air jump after AIR_TIME ms
                else if (uuid.climbCooldown + config.airTime < 0) {
                    player.stopClimbing()
                    return@forEach
                }
            }

            val stamina = (-tickDuration * config.staminaRemovePerTick * atWallMultiplier).let { base ->
                player.inventory.toGeary()?.getEquipmentModifiers(base) ?: base
            }

            if (isClimbing) player.addStamina(stamina)

        }
    }

    private fun flipColor(bar: BossBar) {
        when (BossBar.Color.RED) {
            bar.color() -> {
                bar.color(config.baseBarColor)
                bar.name(baseBar)
            }

            else -> {
                bar.color(BossBar.Color.RED)
                bar.name(redBar) //Make Stamina title red
            }
        }
        bar.overlay(config.baseOverlay)
    }
}
